### PR TITLE
Delete architecture-specific memchr code in std::sys

### PR DESCRIPTION
**Pull request by krtab**
Currently all architecture-specific memchr code is only used in `std::io`. Most of the actual `memchr` capacity exposed to the user through the slice API is instead implemented in `core::slice::memchr`.

Hence this commit deletes `memchr` from `std::sys[_common]` and replace calls to it by calls to `core::slice::memchr` functions. This deletes `(r)memchr` from the list of symbols linked to libc.

The interest of putting architecture specific code back in core is linked to the discussion to be had in #113654

### CODE REVIEW

**Comment by workingjubilee** [comment]
Hmm, I'm assuming this isn't going to affect anything but I would like to see the perf queue say "yes" first.

@bors try @rust-timer queue

**Comment by workingjubilee** [comment]
Hm. Sure. Don't know whether the other changes should be done, but the shocking lack of runtime perf changes feels motivating.

@bors r+

**Comment by workingjubilee** [comment]
uh?
@bors retry

**Comment by krtab** [comment]
Urgh,
A quick morning glance tells me that this is because WASI is lacking memchr, I'll see if I can add polyfills.

**Comment by krtab** [comment]
Turns out it was just that the `sys::memchr` import still existed in the wasi target. The diff since you last reviewed is:

```diff
diff --git a/library/std/src/sys/wasi/os.rs b/library/std/src/sys/wasi/os.rs
index d53bddd8e9d..330c0f4717a 100644
--- a/library/std/src/sys/wasi/os.rs
+++ b/library/std/src/sys/wasi/os.rs
@@ -8,9 +8,9 @@
 use crate::ops::Drop;
 use crate::os::wasi::prelude::*;
 use crate::path::{self, PathBuf};
+use crate::slice::memchr;
 use crate::str;
 use crate::sys::common::small_c_string::{run_path_with_cstr, run_with_cstr};
-use crate::sys::memchr;
 use crate::sys::unsupported;
 use crate::vec;
 ```
 
 (I also rebased, sorry if that messes up the review flow)

**Comment by workingjubilee** [comment]
It's fine, I generally just re-skim the entire thing on every commit.

@bors r+ rollup=iffy

**Comment by workingjubilee** [comment]
erm?

**Comment by workingjubilee** [comment]
damn my lying eyes?

**Comment by ChrisDenton** [comment]
wasm32, library/std/src/sys/wasi/os.rs:13:5. Unresolved import `crate::sys::memchr`, consider importing this module instead `core::slice::memchr`, etc.

**Comment by krtab** [comment]
My bad. I had forgotten the actual change and only committed the rebase. My change also turned out to be slightly wrong but it should be fine now (couldn't get `./x.py check --target wasm32-wasi` to fully pass but the error is not memchr related anymore).

**Comment by workingjubilee** [comment]
Hmm.

@bors r+

**Comment by workingjubilee** [comment]
I'm going to just hand-audit this for every supported platform before I approve this again. Might be a few days before I get around to it.

**Comment by krtab** [comment]
Ok. I'm on break for the next two weeks anyway so no rush.

File: library/std/src/sys/pal/hermit/memchr.rs
Comment (joboet) [inline library/std/src/sys/pal/hermit/memchr.rs:1]: Hermit's `mod.rs` still includes a reference to this module.

File: library/std/src/sys/pal/hermit/os.rs [hunk 1]
```diff
@@ -9,7 +9,6 @@ use crate::os::hermit::ffi::OsStringExt;
 use crate::path::{self, PathBuf};
 use crate::str;
 use crate::sync::Mutex;
-use crate::sys::memchr;
```
Comment (joboet) [inline library/std/src/sys/pal/hermit/os.rs:12]: This is used in line 100.

File: library/std/src/sys/pal/solid/mod.rs
Comment (joboet) [inline library/std/src/sys/pal/solid/mod.rs:1]: `solid/os.rs` has another import.

**Review by joboet — CHANGES_REQUESTED** [review]
There are still some unresolved imports and missing modules on some targets:
* Hermit (x86_64-unknown-hermit)
* SOLID (aarch64-kmc-solid_asp3)
* TeeOS (aarch64-unknown-teeos)
* UEFI still has a `memchr` module (x86_64-unknown-uefi)
* ZKVM does too (it doesn't currently build anyway)

I'd recommend searching all of `std` for `memchr`, checking each use, and then running `x check library/std --target [TARGET]` for each of the targets above, they should be representative.

**Comment by krtab** [comment]
Thanks @joboet, it should hopefully be good know.
And thanks to @WaffleLapkin on zulip as well.

**Comment by workingjubilee** [comment]
Thank you everyone! Sorry for not getting back to you on answering your question on Zulip, the answer was a somewhat inane "manually comb through everything and run this code on the machines I have available".

@bors r+

**Comment by lqd** [comment]
The regression doesn’t look real, syn-opt was somehow less noisy than usual recently, and seems to be back to its usual behavior.

**Pull request merged**